src/jit/machine.rs

```diff
@@ -0,0 +1,726 @@
+/// introduced here for profiling use.
+/// * For very hot code, we might introduce a more sophisticated JIT to
+/// further optimize those code pieces to further boost the performance. In
+/// this level we would leverage algorithms to translate RISC-V instructions
+/// to SSA form: http://compilers.cs.uni-saarland.de/papers/bbhlmz13cc.pdf, then
+/// apply different optimizations to further optimize the code. We might choose
+/// to leverage cranelift or MJIT to enjoy existing work. Note that unlike
+/// the above baseline JIT, this path still has many uncertainties which is more
+/// likely to change. Also this will has much lower priority if baseline JIT
+/// is proved to be enough for CKB use.
+pub struct BaselineJitMachine<'a> {
+    asm_data: AsmData,
+    rust_data: RustData,
+    // In fact program should not belong here, however we are putting it here
+    // so as to shape the API in a way that one instance here only works on
+    // one program
+    program: &'a [u8],
+}
+
+impl<'a> BaselineJitMachine<'a> {
+    pub fn new(program: &'a [u8], tracer: Box<Tracer>) -> Self {
+        Self {
+            asm_data: AsmData::default(),
+            rust_data: RustData::new(tracer),
+            program,
+        }
+    }
+
+    pub fn run(self, args: &[Vec<u8>]) -> Result<(u8, Self), Error> {
+        self.run_with_data(args, BaselineJitRunData::default())
+    }
+
+    pub fn run_with_data<'b>(
+        mut self,
+        args: &[Vec<u8>],
+        data: BaselineJitRunData<'a, 'b>,
+    ) -> Result<(u8, Self), Error> {
+        self.reset();
+        self.rust_data.max_cycles = data.max_cycles;
+        let program = self.program;
+        let mut builder = DefaultMachineBuilder::<Self>::new(self);
+        if let Some(instruction_cycle_func) = data.instruction_cycle_func {
+            builder = builder.instruction_cycle_func(instruction_cycle_func);
+        }
+        for syscall in data.syscalls {
+            builder = builder.syscall(syscall);
+        }
+        let mut machine = builder.build().load_program(program, args)?;
+        let mut emitter = Emitter::new()?;
+        let decoder = build_imac_decoder::<u64>();
+        machine.set_running(true);
+        while machine.running() {
+            let pc = machine.pc().to_usize();
+            let jitted_data = {
+                let rust_data = &mut machine.inner_mut().rust_data;
+                rust_data.tracer.trace(pc)?;
+                if let Some((address, cycles)) = rust_data.jitted_blocks.get(&pc) {
+                    Some((*address, *cycles))
+                } else {
+                    None
+                }
+            };
+            if let Some((address, cycles)) = jitted_data {
+                machine.add_cycles(cycles)?;
+                let core = machine.inner_mut();
+                let rust_data = &mut core.rust_data;
+                if let Some(buffer) = &rust_data.buffer {
+                    let f: fn(&mut AsmData) =
```

Comment:
This is how JIT kicks in here: so whenever we are JITTing a basic block, we translate the block into a function which accepts one `AsmData` struct, which is also [AsmMachine](https://github.com/nervosnetwork/ckb-vm/blob/baseline-jit/src/jit/asm.x64.c#L104-L107) struct at C side. The JITTed assembly code uses this struct to update machine values.

Whenever a new block is translated, we put it in `buffer` and keep a reference to its initial offset(kept in `jitted_blocks`), now when we are executing the basic block here, we first look for the offset of the starting address of the translated function in buffer, cast that back to a function pointer, then pass in the AsmData representing current machine.

It might appear that this parameter is unused, but actually we did use it at assembly side: per x86_64 calling convention, the first parameter in a function is stored in `rdi` register, and we actually use this register [here](https://github.com/nervosnetwork/ckb-vm/blob/baseline-jit/src/jit/asm.x64.c#L288). After that line, we moved `AsmData` to `rbp` register, since `rbp` is persisted across function call, so we would need less save/restore operations than keeping it in `rdi`. After that, we just reference `AsmData` in `rbp` register. That's how we use it.

Reply:
Thanks for the detailed information, if I understand correctly, is it equivalent to?
```
let f = unsafe { mem::transmute::<&u8, fn(&mut AsmData)>(&buffer.as_ref()[address]) };
```

Reply:
Yes I just tried that and I believe it works, thanks for this 👍